apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/profile/ProfileTaskExecutionContext.java

```diff
@@ -72,39 +72,57 @@ public void stopProfiling() {
      *
      * @return is add profile success
      */
-    public ProfileStatusReference attemptProfiling(TracingContext tracingContext,
-                                                   String traceSegmentId,
-                                                   String firstSpanOPName) {
+    public ProfileStatusContext attemptProfiling(TracingContext tracingContext,
+                                                 String traceSegmentId,
+                                                 String firstSpanOPName) {
         // check has available slot
         final int usingSlotCount = currentProfilingCount.get();
         if (usingSlotCount >= Config.Profile.MAX_PARALLEL) {
-            return ProfileStatusReference.createWithNone();
+            return ProfileStatusContext.createWithNone();
         }
 
         // check first operation name matches
         if (!Objects.equals(task.getFirstSpanOPName(), firstSpanOPName)) {
-            return ProfileStatusReference.createWithNone();
+            return ProfileStatusContext.createWithNone();
         }
 
         // if out limit started profiling count then stop add profiling
         if (totalStartedProfilingCount.get() > task.getMaxSamplingCount()) {
-            return ProfileStatusReference.createWithNone();
+            return ProfileStatusContext.createWithNone();
         }
 
         // try to occupy slot
         if (!currentProfilingCount.compareAndSet(usingSlotCount, usingSlotCount + 1)) {
-            return ProfileStatusReference.createWithNone();
+            return ProfileStatusContext.createWithNone();
         }
 
+        ThreadProfiler profiler;
+        if ((profiler = addProfilingThread(tracingContext, traceSegmentId)) != null) {
+            return profiler.profilingStatus();
+        }
+        return ProfileStatusContext.createWithNone();
+    }
+
+    public boolean continueProfiling(TracingContext tracingContext, String traceSegmentId) {
+        // must smaller than max parallels monitoring count
+        if (currentProfilingCount.addAndGet(1) >= Config.Profile.MAX_PARALLEL) {
+            currentProfilingCount.addAndGet(-1);
+            return false;
+        }
+
+        return addProfilingThread(tracingContext, traceSegmentId) != null;
+    }
```

Comment:
So, we still have the threshold to control this for limiting the payload, right?

What is the threshold for UI, and what is for agent by default? Are they reasonably matched?

Reply:
> What is the threshold for UI?

For now, the UI only limits the count of the first segment. The sampling count is not verified when continue profiling the cross-thread segment. 
I think the user only should only care about the profiling limit count of the first segment. The cross-thread segment should be kept profiling. 

> what is for agent by default?

The agent doesn't provide the sampling count, only receives the sampling count from OAP. At present, the agent only provides `profile.max_parallel` to limit the number of agent parallel analyses, so as to prevent dump data from affecting the execution of the business system.

> Are they reasonably matched?

The only thing that may be worried is that the value of `profile.max_parallel` is too small, which may cause the limited cross-thread not to be collected. However, this trade-off point, if the `profile.max_parallel` data is too large, may affect the execution of the business during profiling. If it is too small, it may lead to a lack of collected data.

Reply:
I just want to make sure `profile.max_parallel` is big enough for most cases. Such as for undertow(a typical requirement for this new feature).

Reply:
`Undertow` may be an individual case, because it main thread quickly completes execution and closes, and the sub-thread will execute according to business time. In this case, the `profile.max_parallel` may be enough, because the main-thread slot will empty quickly. 
However, for the usual cross-thread execution, it may not be enough, because the main thread may continue to wait for the sub-thread execution results and return.
So I think maybe we can increase the default value of `profile.max_parallel`. Assuming that a thread will generate an additional sub-thread by default, the existing `5` can be raised to `10`.

Reply:
> So I think maybe we can increase the default value of profile.max_parallel. Assuming that a thread will generate an additional sub-thread by default, the existing 5 can be raised to 10.

By adding this, there are some side effects, such as adding a payload of profiling to sync-process cases.
How about adding a threshold `profile.accept_max_sub_parallel_num`? This should only control the number of sub-tasks.

Reply:
Sure, but is this mean the total sub-thread count for all segments or one segment?

Reply:
I think per-endpoint-access(one main segment)

Reply:
Sure, so this configuration should validate the profiling count of one endpoint(first segment), and the default value could be `1` I think. That's means we only keep one sub-thread profiling.

Reply:
It could be 5, as mostly it should not be in that case.

Reply:
After all releases(new agent and new OAP/UI), we could write a blog introducing this with undertow and real parallel cases. The the default as 5 would make the demo more easier.

Reply:
Updated.